### PR TITLE
feat: cli: New commands for testing.

### PR DESCRIPTION
**Pull request by geoff-vball**
## Proposed Changes
Implements 3 new cli commands.
`compact-paritions` for general use. We never had a way to call this before.
`send-invalid-windowed-post` and `generate-and-send-consensus-fault` for testing purposes.

## Checklist

Before you mark the PR ready for review, please make sure that:
- [ ] All commits have a clear commit message.
- [ ] The PR title is in the form of of `<PR type>: <area>: <change being made>`
    - example: ` fix: mempool: Introduce a cache for valid signatures`
    - `PR type`: _fix_, _feat_, _INTERFACE BREAKING CHANGE_, _CONSENSUS BREAKING_, _build_, _chore_, _ci_, _docs_,_perf_, _refactor_, _revert_, _style_, _test_
    - `area`: _api_, _chain_, _state_, _vm_, _data transfer_, _market_, _mempool_, _message_, _block production_, _multisig_, _networking_, _paychan_, _proving_, _sealing_, _wallet_, _deps_
- [ ] This PR has tests for new functionality or change in behaviour
- [ ] If new user-facing features are introduced, clear usage guidelines and / or documentation updates should be included in https://lotus.filecoin.io or [Discussion Tutorials.](https://github.com/filecoin-project/lotus/discussions/categories/tutorials)
- [ ] CI is green

### CODE REVIEW

File: cmd/lotus-miner/sectors.go [hunk 1]
```diff
@@ -2089,3 +2090,91 @@ func yesno(b bool) string {
...
+			Required: true,
+		},
+		&cli.BoolFlag{
+			Name:  "really-do-it",
+			Usage: "Actually send transaction performing the action",
+			Value: false,
+		},
+		&cli.StringFlag{
+			Name:  "actor",
+			Usage: "TODO",
```
Comment (arajasek) [inline cmd/lotus-miner/sectors.go:2115]: fill TODO

File: cmd/lotus-miner/sectors.go [hunk 2]
```diff
@@ -2089,3 +2090,91 @@ func yesno(b bool) string {
...
+		if err != nil {
+			return err
+		}
+
+		minfo, err := api.StateMinerInfo(ctx, maddr, types.EmptyTSK)
+		if err != nil {
+			return err
+		}
+
+		deadline := cctx.Uint64("deadline")
```
Comment (arajasek) [inline cmd/lotus-miner/sectors.go:2142]: Might as well sanity check `deadline` (within 0 and 47 or whatever the range is)

File: cmd/lotus-miner/sectors.go [hunk 3]
```diff
@@ -2089,3 +2090,91 @@ func yesno(b bool) string {
...
+
+		parts := cctx.Int64Slice("partitions")
+		if len(parts) <= 0 {
+			return fmt.Errorf("must include at least one partition to compact")
+		}
+		fmt.Printf("compacting %d paritions\n", len(parts))
+
+		partitions := bitfield.New()
```
Comment (arajasek) [inline cmd/lotus-miner/sectors.go:2150]: bitfield.NewFromSet is your friend here, i think

File: cmd/lotus-miner/sectors.go [hunk 4]
```diff
@@ -2089,3 +2090,91 @@ func yesno(b bool) string {
...
+			Value:  big.Zero(),
+			Params: sp,
+		}, nil)
+		if err != nil {
+			return xerrors.Errorf("mpool push: %w", err)
+		}
+
+		fmt.Println("Message CID:", smsg.Cid())
+
+		return nil
```
Comment (arajasek) [inline cmd/lotus-miner/sectors.go:2178]: Let's wait for inclusion and report on message success / failure. (see commands like withdraw in lotus-miner/actor.go for example)

File: cmd/lotus-shed/miner.go [hunk 5]
```diff
@@ -354,3 +361,220 @@ var minerUnpackInfoCmd = &cli.Command{
...
+	Description: `Note: This is meant for testing purposes and should NOT be used on mainnet or you will be slashed`,
+	Flags: []cli.Flag{
+		&cli.BoolFlag{
+			Name:  "really-do-it",
+			Usage: "Actually send transaction performing the action",
+			Value: false,
+		},
+		&cli.Int64SliceFlag{
+			Name:     "partitions",
+			Usage:    "list of partitions to compact sectors in",
```
Comment (arajasek) [inline cmd/lotus-shed/miner.go:377]: ?

File: cmd/lotus-shed/miner.go [hunk 6]
```diff
@@ -354,3 +361,220 @@ var minerUnpackInfoCmd = &cli.Command{
...
 	},
 }
+
+var sendInvalidWindowPoStCmd = &cli.Command{
+	Name:        "send-invalid-windowed-post",
+	Usage:       "Sends an invalid windowed post for a specific deadline",
+	Description: `Note: This is meant for testing purposes and should NOT be used on mainnet or you will be slashed`,
+	Flags: []cli.Flag{
+		&cli.BoolFlag{
+			Name:  "really-do-it",
```
Comment (arajasek) [inline cmd/lotus-shed/miner.go:371]: i don't think shed commands need really-do-its, but also no harm in having it...

File: cmd/lotus-shed/miner.go [hunk 7]
```diff
@@ -354,3 +361,220 @@ var minerUnpackInfoCmd = &cli.Command{
...
+		}
+
+		deadline, err := api.StateMinerProvingDeadline(ctx, maddr, types.EmptyTSK)
+		if err != nil {
+			return xerrors.Errorf("getting deadline: %w", err)
+		}
+
+		partitionIndices := cctx.Int64Slice("partitions")
+		if len(partitionIndices) <= 0 {
+			return fmt.Errorf("must include at least one partition to compact")
```
Comment (arajasek) [inline cmd/lotus-shed/miner.go:415]: also ?

File: cmd/lotus-shed/miner.go [hunk 8]
```diff
@@ -354,3 +361,220 @@ var minerUnpackInfoCmd = &cli.Command{
...
+			Params: sp,
+		}, nil)
+		if err != nil {
+			return xerrors.Errorf("mpool push: %w", err)
+		}
+
+		fmt.Println("Message CID:", smsg.Cid())
+
+		return nil
+	},
```
Comment (arajasek) [inline cmd/lotus-shed/miner.go:475]: Would also wait for inclusion and success here -- if you're running this command, you're definitely gonna wanna make sure it succeeds, in which case you might as well do it as part of the command.

File: cmd/lotus-shed/miner.go [hunk 9]
```diff
@@ -354,3 +361,220 @@ var minerUnpackInfoCmd = &cli.Command{
...
+	Description: `Note: This is meant for testing purposes and should NOT be used on mainnet or you will be slashed`,
+	Flags: []cli.Flag{
+		&cli.BoolFlag{
+			Name:  "really-do-it",
+			Usage: "Actually send transaction performing the action",
+			Value: false,
+		},
+		&cli.StringFlag{
+			Name:  "actor",
+			Usage: "TODO",
```
Comment (arajasek) [inline cmd/lotus-shed/miner.go:490]: You don't really need this -- you can figure out the miner from the block being inputted.

File: cmd/lotus-shed/miner.go [hunk 10]
```diff
@@ -354,3 +361,220 @@ var minerUnpackInfoCmd = &cli.Command{
...
+		},
+	},
+	Action: func(cctx *cli.Context) error {
+		if !cctx.Bool("really-do-it") {
+			return xerrors.Errorf("Pass --really-do-it to actually execute this action")
+		}
+
+		if cctx.Args().Len() != 1 {
```
Comment (arajasek) [inline cmd/lotus-shed/miner.go:498]:
```suggestion
		if cctx.NArgs() != 1 {
```

File: cmd/lotus-shed/miner.go [hunk 11]
```diff
@@ -354,3 +361,220 @@ var minerUnpackInfoCmd = &cli.Command{
...
+			return xerrors.Errorf("getting miner info: %w", err)
+		}
+
+		blockHeader, err := api.ChainGetBlock(ctx, blockCid)
+		if err != nil {
+			return xerrors.Errorf("getting block header: %w", err)
+		}
+
+		blockHeaderCopy := *blockHeader
+		blockHeaderCopy.ForkSignaling = blockHeader.ForkSignaling + 1
```
Comment (arajasek) [inline cmd/lotus-shed/miner.go:531]: Comment above this indicating what it's doing pls

File: cmd/lotus-shed/miner.go [hunk 12]
```diff
@@ -354,3 +361,220 @@ var minerUnpackInfoCmd = &cli.Command{
...
+			Value:  big.Zero(),
+			Params: sp,
+		}, nil)
+		if err != nil {
+			return xerrors.Errorf("mpool push: %w", err)
+		}
+
+		fmt.Println("Message CID:", smsg.Cid())
+
+		return nil
```
Comment (arajasek) [inline cmd/lotus-shed/miner.go:578]: Also wait for inclusion and success here.

**Review by arajasek — COMMENTED** [review]
Nits, nits galore.

File: cmd/lotus-miner/sectors.go (hunk 3 above)
Comment (geoff-vball) [reply cmd/lotus-miner/sectors.go:2150]: Unfortunately, we can only take a []int64 from the cli, and we need to cast each element to uint64 before putting it into the bitfield, and I don't think we can cast the whole slice at once.

**Review by arajasek — APPROVED** [review]

**Pull request merged**